File: dijkstra.go
package grok

import (
	"math"
)

// DRY tests
type conn map[string]int

// dag == directed acyclic graph
type dag map[string]conn

func dijkstra(graph dag) (weightShortestPath int) {
	// assumes "fin" is destination node
	processed := map[string]bool{}
	costs := map[string]int{}
	parents := map[string]string{"fin": ""}

	// populate costs and parents
	for node, conns := range graph {
		for con, cos := range conns {
			if _, ok := costs[con]; !ok {
				costs[con] = cos
				parents[con] = node
			}
		}
	}

	costs["fin"] = math.MaxInt8

	findLowestCostNode := func() string {
		lowestCost := math.MaxInt8
		lowestCostNode := ""

		for node, cost := range costs {
			if cost < lowestCost && !processed[node] {
				lowestCost = cost
				lowestCostNode = node
			}
		}

		return lowestCostNode
	}

	var cost int
	var neighbors conn

	node := findLowestCostNode()

	for node != "" {
		cost = costs[node]
		neighbors = graph[node]

		for nname, ncost := range neighbors {
			newCost := cost + ncost

			if costs[nname] > newCost {
				costs[nname] = newCost
				parents[nname] = node
			}

		}

		processed[node] = true
		node = findLowestCostNode()
	}

	return costs["fin"]
}

File: dijkstra.js
const tap = require("tap");

function dijkstra(graph) {
  const costs = { fin: Infinity };
  const parents = { fin: null };

  for (const g in graph["start"]) {
    costs[g] = graph["start"][g];
    parents[g] = "start";
  }

  const processed = {};

  function findLowestCostNode() {
    let lowestCost = Infinity;
    let lowestCodeNode;

    for (const node in costs) {
      const cost = costs[node];

      if (cost < lowestCost && !processed[node]) {
        lowestCode = cost;
        lowestCodeNode = node;
      }
    }

    return lowestCodeNode;
  }

  let node = findLowestCostNode();

  while (node != null) {
    const cost = costs[node];
    const neighbors = graph[node];

    for (const n in neighbors) {
      const newCost = cost + neighbors[n];

      if (costs[n] > newCost) {
        costs[n] = newCost;
        parents[n] = node;
      }
    }

    processed[node] = true;

    node = findLowestCostNode();
  }

  return costs["fin"];
}

tap.equal(
  dijkstra({
    start: {
      a: 6,
      b: 2
    },
    a: {
      fin: 1
    },
    b: {
      a: 3,
      fin: 5
    },
    fin: {}
  }),
  6
);
